compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker.hs

```diff
@@ -20,36 +19,39 @@ import qualified DA.Daml.LF.TypeChecker.PartyLiterals as PartyLits
 import qualified DA.Daml.LF.TypeChecker.Recursion as Recursion
 import qualified DA.Daml.LF.TypeChecker.Serializability as Serializability
 import qualified DA.Daml.LF.TypeChecker.NameCollision as NameCollision
+import Development.IDE.Types.Diagnostics
 
 checkModule ::
      World
   -> Version
   -> Module
-  -> Either Error ()
+  -> [Diagnostic]
 checkModule world0 version m = do
   checkModuleInWorld (extendWorldSelf m world0) version m
 
 checkPackage ::
      World
   -> Version
-  -> Either Error ()
-checkPackage world version = do
-  let package = getWorldSelf world
-  let modules = NM.toList (packageModules package)
-  forM_ modules $ \m -> do
-    checkModuleInWorld world version m
+  -> [Diagnostic]
+checkPackage world version = concatMap (checkModuleInWorld world version) modules
+    where
+      package = getWorldSelf world
+      modules = NM.toList (packageModules package)
 
-checkModuleInWorld :: World -> Version -> Module -> Either Error ()
-checkModuleInWorld world version m = do
-    runGamma world version $ do
-      -- We must call `Recursion.checkModule` before `Check.checkModule`
-      -- or else we might loop, attempting to expand recursive type synonyms
-      Recursion.checkModule m
-      Check.checkModule m
-      Serializability.checkModule m
-      PartyLits.checkModule m
-    NameCollision.runCheckModuleDeps world m
+checkModuleInWorld :: World -> Version -> Module -> [Diagnostic]
+checkModuleInWorld world version m =
+    case typeCheckResult of
+        Left err -> [toDiagnostic DsError err]
+        Right () -> NameCollision.runCheckModuleDeps world m
```

Comment:
Would it make sense to deliver the errors and the warnings in one go, i.e., append the diagnostics rather than branch here?

Reply:
Changed, not sure how useful it is in practice given that things are pretty bad if the LF typechecker fails but it doesn’t do any harm.